btrfs-progs: device_get_partition_size*() enhancement and cleanup

adam900710 opened the pull request:
[CHANGELOG]
v2:
- Include and update Zoltan's fix to device_get_partition_size_sysfs()
  The "/dev/block/<device>/size" attribute is always in sector unit
  (512Bytes), independent from the physical device's block size.

  So the fix is pretty straightforward, just do the left shift before
  returning.

  And update the commit message to properly reflect that.

- Use s64 as the return value of device_get_partition_size*()
  Kernel ensures the max sector number of a device is LLONG_MAX >> 9,
  so s64 is more than enough to cover the device size, and we are safe
  to use the minus part to pass errors.

- Extra error handling when device_get_partition_size*() failed
  Ouput an error message and exit.
  Most callers are already doing that correctly but some are not.

- Keep the sysfs method as the fallback way to grab device size
  Since the device_get_partition_size_sysfs() can be easily fixed, no
  need to use complex path search for sector size.

- Add a new cleanup to remove is_vol_small()

Zoltan recently sent a fix to solve the bug in
device_get_partition_size_sysfs(), which is causing problems for "btrfs
dev usage".

It turns out that, the bug is just the attribute
"/sys/block/<device>/size" is in sector unit (512B), not in bytes.

So fix the bug first by reporting the correct size in bytes.

Then remove several unused functions exposed during the error handling
cleanup.

Finally cleanup device_get_partition_size*() helpers to provide a proper
error handling, not just return 0 for errors, but proper minus erro
code, and extra overflow check.

raczzoli committed:
btrfs-progs: fix the wrong size from device_get_partition_size_sysfs()

[BUG]
When an unprivileged user, who can not access the block device, run
"btrfs dev usage", it's very common to result the following incorrect
output:

  $ btrfs dev usage /mnt/btrfs/
  WARNING: cannot read detailed chunk info, per-device usage will not be shown, run as root
  /dev/mapper/test-scratch1, ID: 1
     Device size:            20.00MiB <<<
     Device slack:           16.00EiB <<<
     Unallocated:                 N/A

Note if the unprivileged user has read access to the raw block file, it
will work as expected:

  $ btrfs dev usage /mnt/btrfs/
  WARNING: cannot read detailed chunk info, per-device usage will not be shown, run as root
  /dev/mapper/test-scratch1, ID: 1
     Device size:            10.00GiB
     Device slack:              0.00B
     Unallocated:                 N/A

[CAUSE]
When device_get_partition_size() is called, firstly the function checks
if we can do a read-only open() on the block device.

However under most distros, block devices are only accessible by root
and "disk" group.

If the unprivileged user is not in "disk" group, the open() will fail
and we have to fallback to device_get_partition_size_sysfs() as the
fallback.

The function device_get_partition_size_sysfs() will use
"/sys/block/<device>/size" as the size of the disk.

But according to the kernel source code, the "size" attribute is
implemented by returning bdev_nr_sectors(), and that result is always in
sector unit (512 bytes).

So if device_get_partition_size_sysfs() returns the value directly, it's
512 times smaller than the original size, causing errors.

[FIX]
Just do the proper left shift to return size in bytes.

Issue: #979
Signed-off-by: Qu Wenruo <wqu@suse.com>

adam900710 committed:
btrfs-progs: remove the unnecessary device_get_partition_size() call

Inside function _cmd_filesystem_usage_tabular(), there is a call of
device_get_partition_size() to calculate @unused.

However immediately after that call, @unused is updated using
devinfo->size, so that the previous call of device_get_partition_size()
is completely useless.

Just remove the unnecessary call.

And since we're here, also remove a dead newline in that loop at
variable declaration.

Fixes: 7ab0c46a814a ("btrfs-progs: fi usage: fix unallocated and print total and slack")
Reviewed-by: Anand Jain <anand.jain@oracle.com>
Signed-off-by: Qu Wenruo <wqu@suse.com>

adam900710 committed:
btrfs-progs: remove device_get_partition_size_fd()

The last user of the helper is removed in commit 585ac14d1af2
("btrfs-progs: use btrfs_device_size() instead of
device_get_partition_size_fd()"), and that helper is not generic enough
to handle regular files.

So let's just remove it.

Reviewed-by: Anand Jain <anand.jain@oracle.com>
Signed-off-by: Qu Wenruo <wqu@suse.com>

kdave commented:
> And the original bug is that the sysfs interface only gives the device size in block unit, and block size can be hard to fetch (partition and dm have different path to grab the block size).
> 
> This makes me wonder, why we need the sysfs interface in the first place. If the unprivileged user has no access to a block device, we should not workaround it, but give a proper error/warning message.

The sysfs fallback is for the case when the device size is not readable from the block device but the size is not a secret information. Eg. it's also visible in /proc/partitions, there's no reason to hide it from the user. Forcing to use sudo for basic size information is going against usability.

Reporting the device slack needs some additional information so this should be fixed but otherwise reading the device sizes should stay as is.

adam900710 commented:
OK, I'll drop the last patch so the sysfs fallback is still available, and allow Zotan's fix to be applied.

adam900710 committed:
btrfs-progs: add error handling for device_get_partition_size()

The function device_get_partition_size() has all kinds of error paths,
but it has no way to return error other than returning 0.

This is not helpful for end users to know what's going wrong.

Change that function to return int for error code, and a dedicated u64
pointer to return the device size in bytes.

All existing call sites will exit early using that error code.
The only exception is load_device_info() which will only give a warning
and continue.

Signed-off-by: Qu Wenruo <wqu@suse.com>

adam900710 committed:
btrfs-progs: remove is_vol_small() helper

The last user of the helper is removed in commit c11e36a29e84
("Btrfs-progs: Do not force mixed block group creation unless '-M'
option is specified").

So there is no one using this helper, and we can safely remove it.

Reviewed-by: Anand Jain <anand.jain@oracle.com>
Signed-off-by: Qu Wenruo <wqu@suse.com>

adam900710 committed:
btrfs-progs: add error handling for device_get_partition_size_fd_stat()

The function device_get_partition_size_fd_state() has two different
error paths:

- The target file is not a regular nor block file
  This should be the common one.

- ioctl failed
  This should be very rare.

But it has no way to return error other than returning 0.
This is not helpful for end users to know what's going wrong.

Change that function to return int, and a u64 pointer to return the size
in bytes.

And since we're here, also enhance the error handling of the callers to
do an explicit error message and return error or exit.

Signed-off-by: Qu Wenruo <wqu@suse.com>

kdave reviewed (approved):
Ok with the s64 change discussed in the mailinglist. Thanks.
